### PR TITLE
Consolidating new_<>()

### PR DESCRIPTION
**Pull request by hkaiser**
This patch introduces a consolidation of `hpx::new_<>()`. The following overloads have been implemented:

    // create single instances of a component
    future<id_type> f = new_<Component>(loc, ...);
    client c = new_<client>(loc, ...);

    // create multiple instances of a component
    future<std::vector<id_type>> f = new_<Component[]>(loc, ...);
    future<std::vector<client>> f = new_<client[]>(loc, ...);

This patch also introduces 3 distribution policies: `default_distribution_policy`, `binpacking_distribution_policy`, and `collocating_distribution_policy`. All overloads of `new_` above can be invoked with one of the distribution policies instead of the target locality `loc`, for instance:

     future<id_type> f = new_<Component>(colocated(id), ...);

All of the above also allowed to remove the `new_colocated` API functions, and the `distributing_factory` and `binpacking_factory` components.

For convenience, this patch also renames `HPX_REGISTER_MINIMAL_COMPONENT_FACTORY` to `HPX_REGISTER_COMPONENT`.

### CODE REVIEW

**Comment by hkaiser** [comment]
@sithhell: The last commit fixed the things you mentioned.

File: hpx/runtime/components/binpacking_distribution_policy.hpp [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+            // handle special cases
+            if (localities_.size() == 0)
+            {
+                return stub_base<Component>::create_async(
+                    hpx::find_here(), std::forward<Ts>(vs)...);
+            }
+            else if (localities_.size() == 1)
+            {
+                return stub_base<Component>::create_async(
+                    localities_.front(), std::forward<Ts>(vs)...);
```
Comment (sithhell) [inline hpx/runtime/components/binpacking_distribution_policy.hpp]: Is it intended to only use the first locality here? Shouldn't it rather use the "best"?

Comment (hkaiser) [inline hpx/runtime/components/binpacking_distribution_policy.hpp]: If there is only one locality it's always the first one, no?

Comment (sithhell) [inline hpx/runtime/components/binpacking_distribution_policy.hpp]: D'oh...

**Pull request merged**